### PR TITLE
Add Camellia CBC CTS mode support to EVP layer

### LINKED ISSUE #16276: Add Camellia CTS implementation to default provider
We are missing Camellia CTS in the high level API which was possible to do with the old APIs.
See also #16235

### PR DESCRIPTION
**Pull request by slontis**
Fixes #16276

The provider related AES CTS code has been renamed so that it can be
shared with Camellia. The only real difference is the underlying cbc
cipher that is used (which is pointed to in a generic manner via the hw
pointer).

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
No tests or documentation yet..

File: providers/implementations/ciphers/cipher_cts.c [hunk 1]
```diff
@@ -48,19 +48,20 @@
...
 #include "internal/nelem.h"
-#include "cipher_aes_cts.h"
+#include "cipher_cts.h"
 
 /* The value assigned to 0 is the default */
 #define CTS_CS1 0
 #define CTS_CS2 1
 #define CTS_CS3 2
 
+#define CTS_BLOCK_SIZE 16
```
Comment (t8m) [inline providers/implementations/ciphers/cipher_cts.c:60]: This effectively limits this generic CTS implementation to block ciphers with 128 bit block size. Maybe add a comment above.

File: providers/implementations/ciphers/cipher_cts.c [hunk 2]
```diff
@@ -8,10 +8,10 @@
  */
 
 /*
- * Helper functions for AES CBC CTS ciphers.
+ * Helper functions for AES/CAMELLIA CBC CTS ciphers.
```
Comment (t8m) [inline providers/implementations/ciphers/cipher_cts.c:11]: Perhaps - `Helper functions for 128bit block size CBC CTS ciphers (currently AES/CAMELLIA)`

**Comment by slontis** [comment]
The encryption test vectors listed here in
https://mirror.unpad.ac.id/ietf/internet-drafts/draft-ietf-krb-wg-camellia-cts-00.txt
are missing a vital piece of information.. which is listed here.
https://www.rfc-editor.org/errata_search.php?rfc=6803

**Comment by slontis** [comment]
Tests and Documentation added. Commits have been split up.

**Comment by slontis** [comment]
It took longer to derive the test vectors than it did to write the actual code..
The derived key for camellia CTS seems to use KBKDF whereas another RFC's for KRB uses KRB5KDF (sigh!)

**Comment by slontis** [comment]
Even if this doesnt go in for 3.0 there are some bug fixes in this PR that probably should be included..

**Comment by beldmit** [comment]
Nice work!

**Comment by t8m** [comment]
> Even if this doesnt go in for 3.0 there are some bug fixes in this PR that probably should be included..

Could you please split those out into a separate PR? I do not think this can go into 3.0 as a whole.

**Comment by beldmit** [comment]
I'd like to have it in 3.0 (otherwise we possibly will have to apply it as a downstream patch), but I agree that the bugfixes are worth separating if possible.

File: doc/man7/migration_guide.pod [hunk 3]
```diff
@@ -1215,8 +1216,18 @@ CRYPTO_mem_debug_pop(), CRYPTO_mem_debug_push(), CRYPTO_mem_debug_realloc(),
 CRYPTO_mem_leaks(), CRYPTO_mem_leaks_cb(), CRYPTO_mem_leaks_fp(),
 CRYPTO_set_mem_debug()
 
-Memory-leak checking has been deprecated in favor of more modern development
-tools, such as compiler memory and leak sanitizers or Valgrind.
```
Comment (h-vetinari) [inline doc/man7/migration_guide.pod:1219]: This removal seems unrelated...

File: providers/implementations/ciphers/cipher_cts.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "crypto/evp.h"
+
+/* NOTE: The underlying block cipher is AES CBC so we reuse most of the code */
```
Comment (paulidale) [inline providers/implementations/ciphers/cipher_cts.h:12]: Isn't not AES anymore :)

File: doc/man7/migration_guide.pod (hunk 3 above)
Comment (paulidale) [reply doc/man7/migration_guide.pod:1219]: and suspicious.

**Review by paulidale — APPROVED** [review]
Remove section looks suspicious from the migration guide.  Otherwise, just a wording fix in a comment.

Looks good.

File: doc/man7/migration_guide.pod (hunk 3 above)
Comment (slontis) [reply doc/man7/migration_guide.pod]: oops..

**Comment by slontis** [comment]
> > Even if this doesnt go in for 3.0 there are some bug fixes in this PR that probably should be included..
> 
> Could you please split those out into a separate PR? I do not think this can go into 3.0 as a whole.

The PR really has 3 parts:
1) Move existing CTS code into a separate file so that it can be shared. Make AES_CTS use this.
2) A direct copy of the AES code to setup CAMELLIA - the only change really is the names passed to the tables
3) Bug fixes 
  (i) Handle a 16 byte input  for CS3 as normal CBC instead of erroring.
  (ii) the decrypt was not returning the correct updated iv.

**Comment by kaduk** [comment]
> It took longer to derive the test vectors than it did to write the actual code..
> The derived key for camellia CTS seems to use KBKDF whereas another RFC's for KRB uses KRB5KDF (sigh!)

The key-derivation process is rather intrinsic to the RFC 3961 cryptographic suite ... which you should not be expected to know just from the request being made of openssl.  Which KDF to use is a property of the encryption or checksum type, and as we define new encryption/checksum algorithms we try to take advantage of improvements in KDF design as well.

**Comment by beldmit** [comment]
Do we have any decision on this?

**Comment by t-j-h** [comment]
Discussed on the call.
Waiting on OTC vote result (and for the OTC vote text to be emailed out @mattcaswell ).

**Comment by frozencemetery** [comment]
Thanks for making this PR.  It appears to work with krb5 in my [test branch](https://github.com/frozencemetery/krb5/commit/8f79e68fbeaf88fbcd7857e060c9b48f472c3209).

**Comment by paulidale** [comment]
Pending final vote result.

**Comment by t-j-h** [comment]
With @beldmit's vote cast the decision is clear and you can close the vote and proceed imho.

**Comment by paulidale** [comment]
Vote closed and merged.
